packages/server/customer-os-webhooks/service/sync_email_service.go

```diff
@@ -35,47 +44,267 @@ func NewSyncEmailService(log logger.Logger, repositories *repository.Repositorie
+		if err != nil {
+			logrus.Errorf("failed to convert emailData sent date to UTC for emailData with id %v :%v", emailData.Id, err)
+		}
+
+		from, to, cc, bcc, references, inReplyTo := extractEmailData(emailData)
+
+		personalEmailProviderList, err := s.services.CommonServices.CommonRepositories.PersonalEmailProviderRepository.GetPersonalEmailProviders()
+		if err != nil {
+			logrus.Errorf("failed to get personal emailData provider list: %v", err)
+		}
+
+		allEmailsString, err := buildEmailsListExcludingPersonalEmails(personalEmailProviderList, "", emailData.SentBy, to, cc, bcc)
+		if err != nil {
+			logrus.Errorf("failed to build emails list: %v", err)
+		}
+
+		if len(allEmailsString) == 0 {
+			return SyncResult{}, SyncResult{}, SyncResult{}, nil
+		}
+
+		// Create a map to store the domain counts
+		domainCount := make(map[string]int)
+
+		// Iterate through the email addresses
+		for _, email := range allEmailsString {
+			domain := utils.ExtractDomain(email)
+			if domain != "" {
+				domainCount[domain]++
+			}
+		}
+
+		if len(domainCount) > 5 {
+			//reason := "more than 5 domains belongs to a workspace domain"
+			return SyncResult{}, SyncResult{}, SyncResult{}, nil
+		}
+
+		channelData, err := buildEmailChannelData(emailData.Subject, references, inReplyTo)
+		if err != nil {
+			logrus.Errorf("failed to build emailData channel data for emailData with id %v: %v", emailData.Id, err)
+			return SyncResult{Skipped: 1}, SyncResult{}, SyncResult{}, nil
+		}
+
+		sessionId, err := s.services.InteractionSessionService.MergeInteractionSession(ctx, common.GetTenantFromContext(ctx), emailData.ExternalSystem, emailData.SessionDetails, now)
+
+		if err != nil {
+			logrus.Errorf("failed merge interaction session for emailData id %v :%v", emailData.Id, err)
+			return SyncResult{}, SyncResult{}, SyncResult{}, nil
+		}
+
+		//TODO check integration event data mapping
+		integrationEvent := model.InteractionEventData{
+			BaseData:       model.BaseData{CreatedAt: &emailSentDate},
+			Content:        emailData.Content,
+			ContentType:    emailData.ContentType,
+			Channel:        emailData.Channel,
+			ChannelData:    *channelData,
+			Identifier:     emailData.Identifier,
+			EventType:      emailData.EventType,
+			Hide:           emailData.Hide,
+			BelongsTo:      emailData.BelongsTo,
+			SessionDetails: emailData.SessionDetails,
+		}
+		var interactionEvents []model.InteractionEventData
+		interactionEvents = append(interactionEvents, integrationEvent)
+
+		interactionEventSyncResult, err = s.services.InteractionEventService.SyncInteractionEvents(ctx, interactionEvents)
+		if err != nil {
+			logrus.Errorf("failed merge interaction event for emailData id %v :%v", emailData.Id, err)
+			return SyncResult{}, SyncResult{Failed: 1}, SyncResult{}, nil
+		}
+
+		err = s.repositories.Neo4jRepositories.InteractionEventWriteRepository.LinkInteractionEventToSession(ctx, common.GetTenantFromContext(ctx), interactionEventId, sessionId)
```

Comment:
this probably will fail. the code execution here will be fast and will not give time for the interaction session and itneraction event to be actually pushed in neo4j. this will fail or will not link the nodes.

please add checks before to wait for interaction event and interaction session node to pe persisted in db. the timeout should be like a 10 seconds. if the context will expire, the email process will fail so we need to store the actual request in the database or we need a retry mechanism